netmaster/daemon.go

```diff
@@ -0,0 +1,339 @@
+/***
+Copyright 2014 Cisco Systems Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/contiv/netplugin/core"
+	"github.com/contiv/netplugin/drivers"
+	"github.com/contiv/netplugin/netmaster/master"
+	"github.com/contiv/netplugin/netmaster/mastercfg"
+	"github.com/contiv/netplugin/netmaster/objApi"
+	"github.com/contiv/netplugin/utils"
+	"github.com/contiv/objdb"
+	"github.com/contiv/ofnet"
+	"github.com/gorilla/mux"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const leaderLockTTL = 10
+
+type daemon struct {
+	listenURL        string                // URL where netmaster needs to listen
+	currState        string                // Current state of the daemon
+	apiController    *objApi.APIController // API controller for contiv model
+	stateDriver      core.StateDriver      // KV store
+	objdbClient      objdb.API             // Objdb client
+	ofnetMaster      *ofnet.OfnetMaster    // Ofnet master instance
+	listenerMutex    sync.Mutex            // Mutex for HTTP listener
+	stopLeaderChan   chan bool             // Channel to stop the leader listener
+	stopFollowerChan chan bool             // Channel to stop the follower listener
+}
+
+var leaderLock objdb.LockInterface // leader lock
+
+func (d *daemon) registerService() {
+	// Get the address to be used for local communication
+	localIP, err := d.objdbClient.GetLocalAddr()
+	if err != nil {
+		log.Fatalf("Error getting local IP address. Err: %v", err)
+	}
+
+	// service info
+	srvInfo := objdb.ServiceInfo{
+		ServiceName: "netmaster",
+		HostAddr:    localIP,
+		Port:        9999,
+		Role:        d.currState,
+	}
+
+	// Register the node with service registry
+	err = d.objdbClient.RegisterService(srvInfo)
+	if err != nil {
+		log.Fatalf("Error registering service. Err: %v", err)
+	}
+
+	log.Infof("Registered netmaster service with registry")
+}
+
+// registerWebuiHandler registers handlers for serving web UI
+func (d *daemon) registerWebuiHandler(router *mux.Router) {
+	// Setup the router to serve the web UI
+	goPath := os.Getenv("GOPATH")
+	if goPath != "" {
+		webPath := goPath + "/src/github.com/contiv/contivmodel/www/"
+
+		// Make sure we have the web UI files
+		_, err := os.Stat(webPath)
+		if err != nil {
+			webPath = goPath + "/src/github.com/contiv/netplugin/" +
+				"Godeps/_workspace/src/github.com/contiv/contivmodel/www/"
```

Comment:
Yes. Thanks. its on my todo list to make this a command line argument instead of hard coding.
Next PR may be..